src/queue.rs

```diff
@@ -399,70 +398,74 @@ pub struct Queue<M: GuestAddressSpace> {
     pub device_area: GuestAddress,
 }
 
-impl<M: GuestAddressSpace> Queue<M> {
-    /// Constructs an empty virtio queue with the given `max_size`.
-    pub fn new(mem: M, max_size: u16) -> Queue<M> {
+impl QueueConfig {
+    /// Create `QueueConfig` object based on the specified `max_size` and default values for
+    /// the other fields.
+    pub fn with_max_size(max_size: u16) -> Self {
+        QueueConfig {
+            max_size,
+            ..Default::default()
+        }
+    }
+}
+
+/// A virtio split queue.
+#[derive(Clone)]
+pub struct Queue<M, C> {
+    mem: M,
+    config: C,
+}
+
+impl<M: GuestAddressSpace> Queue<M, QueueConfig> {
+    /// Create a new `Queue` that holds a `QueueConfig` object.
+    pub fn new(mem: M, max_size: u16) -> Queue<M, QueueConfig> {
         Queue {
             mem,
-            max_size,
-            size: max_size,
-            ready: false,
-            desc_area: GuestAddress(0),
-            driver_area: GuestAddress(0),
-            device_area: GuestAddress(0),
-            next_avail: Wrapping(0),
```

Comment:
I am not sure that `ready`/`next_avail`/`next_used`/`event_idx_enabled`/`signalled_used` should be part of the `QueueConfig`. They are runtime state that should only be setup at `activate()` time; therefore it may belong in the `Queue`s that `activate()` creates.

Reply:
The config struct (`QueueState` might have been a better name) attempts to capture all the characteristics of a `Queue` object that are independent of the memory handle. Further separating this into pre- and post-activate configuration is def possible, but ultimately we still need a way to reconstruct a `Queue` from the latter (for example, when restoring previously saved state), and it seemed more straightforward to keep everything together.

Reply:
> ultimately we still need a way to reconstruct a Queue from the latter

Ok, that makes sense then. I thought it was just for the configuration as done by the guest, because it is passed to `activate`. Renaming to QueueState would then be a good idea.